File: apps/fsharp/src/MachineLearning/Chapter07/Main.fs
module MachineLearning.Chapter07.Main

open System.IO
open MachineLearning.Dataset
open MachineLearning.Chapter07.Cinema
open MachineLearning.Chapter07.LinearRegression
open MachineLearning.Chapter07.RegressionMetrics
open MachineLearning.Chapter07.StatsRegression

[<Literal>]
let TestSize = 0.2

[<Literal>]
let Seed = 0

let private formatCoefficients (coefficients: Map<string, float>) : string =
    coefficients
    |> Map.toList
    |> List.map (fun (name, value) -> $"{name}={value:F4}")
    |> String.concat ", "

/// cinema.csv で線形回帰を学習し、係数とテストデータの評価、FSharp.Stats との比較を表示する
let run (print: string -> unit) : unit =
    let csvFile = Path.Combine(dataDir (), "cinema.csv")
    let rows = loadCinema csvFile
    let split = prepareCinema csvFile TestSize Seed
    let model = fitLinearRegression split.XTrain split.TTrain
    let y = predictLinearRegression model split.XTest
    let library = fitWithFSharpStats split.XTrain split.TTrain
    let libraryY = predictLinearRegression library split.XTest

    print $"データ件数: {rows.Length}"
    print $"外れ値を除いた件数: {(removeOutliers rows).Length}"
    print $"訓練データ: {split.XTrain.Length} 件, テストデータ: {split.XTest.Length} 件"
    print $"切片: {model.Intercept:F2}"
    print $"係数: {formatCoefficients model.Coefficients}"

    print
        $"テストデータの評価: R2={r2Score split.TTest y:F4}, MAE={meanAbsoluteError split.TTest y:F2}, RMSE={rootMeanSquaredError split.TTest y:F2}"

    print $"FSharp.Stats: 切片={library.Intercept:F2}, 係数: {formatCoefficients library.Coefficients}"
    print $"FSharp.Stats の R2: {r2WithFSharpStats split.TTest libraryY:F4}"

File: apps/scala/src/main/scala/machinelearning/chapter07/Main.scala
package machinelearning.chapter07

import java.nio.file.Paths
import java.util.Locale
import machinelearning.chapter02.Table
import machinelearning.dataset.DataDir

/** 映画の興行収入を線形回帰で予測し、係数と評価指標を表示する。 */
object Main:
  private val TestSize = 0.2
  private val Seed = 0L

  def run(print: String => Unit): Unit =
    val csvFile = Paths.get(DataDir.current(), "cinema.csv")
    val table = Table.load(csvFile)
    val split = Cinema.prepare(csvFile, TestSize, Seed)
    val model = LinearRegression.fit(split.xTrain, split.tTrain)
    val y = model.predict(split.xTest)
    val coefficients =
      model.coefficients.map((name, value) => s"$name=${format(value, 4)}").mkString(", ")
    print(s"データ件数: ${table.rows.size}")
    print(s"外れ値を除いた件数: ${Cinema.removeOutliers(table).rows.size}")
    print(s"訓練データ: ${split.xTrain.size} 件, テストデータ: ${split.xTest.size} 件")
    print(s"切片: ${format(model.intercept, 2)}")
    print(s"係数: $coefficients")
    print(
      "テストデータの評価: " +
        s"R2=${format(RegressionMetrics.r2Score(split.tTest, y), 4)}, " +
        s"MAE=${format(RegressionMetrics.meanAbsoluteError(split.tTest, y), 2)}, " +
        s"RMSE=${format(RegressionMetrics.rootMeanSquaredError(split.tTest, y), 2)}"
    )

  private def format(value: Double, decimals: Int): String =
    String.format(Locale.ROOT, s"%.${decimals}f", value)
